src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -319,90 +328,48 @@ private unsafe int IndexOfOrdinalIgnoreCaseHelper(ReadOnlySpan<char> source, Rea
-                        if ((uint)(valueChar - 'a') <= ('z' - 'a'))
-                            valueChar = (char)(valueChar - 0x20);
-                        if ((uint)(targetChar - 'a') <= ('z' - 'a'))
-                            targetChar = (char)(targetChar - 0x20);
-
-                        if (valueChar >= 0x80 || s_highCharTable[valueChar])
-                            goto InteropCall;
-                        else if (valueChar != targetChar)
-                            break;
-                        sourceIndex++;
-                    }
-
-                    if (targetIndex == target.Length)
-                    {
-                        if (matchLengthPtr != null)
-                            *matchLengthPtr = target.Length;
-                        return i;
-                    }
+                    startIndex = 0;
+                    endIndex = source.Length - value.Length + 1;
+                    jump = 1;
                 }
-                if (i > endIndex)
-                    return -1;
-            InteropCall:
-                return Interop.Globalization.IndexOf(_sortHandle, b, target.Length, a, source.Length, options, matchLengthPtr);
-            }
-        }
-
-        private unsafe int IndexOfOrdinalHelper(ReadOnlySpan<char> source, ReadOnlySpan<char> target, CompareOptions options, int* matchLengthPtr)
-        {
-            Debug.Assert(!_invariantMode);
-
-            Debug.Assert(!source.IsEmpty);
-            Debug.Assert(!target.IsEmpty);
-            Debug.Assert(_isAsciiEqualityOrdinal);
-
-            fixed (char* ap = &MemoryMarshal.GetReference(source))
-            fixed (char* bp = &MemoryMarshal.GetReference(target))
-            {
-                char* a = ap;
-                char* b = bp;
-                int endIndex = source.Length - target.Length;
-
-                if (endIndex < 0)
-                    goto InteropCall;
-
-                for (int j = 0; j < target.Length; j++)
+                else
                 {
-                    char targetChar = *(b + j);
-                    if (targetChar >= 0x80 || s_highCharTable[targetChar])
-                        goto InteropCall;
+                    startIndex = source.Length - value.Length;
+                    endIndex = -1;
+                    jump = -1;
                 }
 
-                int i = 0;
-                for (; i <= endIndex; i++)
+                for (int i = startIndex; i != endIndex; i += jump)
                 {
                     int targetIndex = 0;
                     int sourceIndex = i;
 
-                    for (; targetIndex < target.Length; targetIndex++)
+                    for (; targetIndex < target.Length; targetIndex++, sourceIndex++)
                     {
                         char valueChar = *(a + sourceIndex);
                         char targetChar = *(b + targetIndex);
+
+                        if (ignoreCase)
```

Comment:
>if (ignoreCase) [](start = 24, length = 15)

I am not sure if the JIT will optimize this check or it will be checked with every iteration in the loop? 

Would it be better to duplicate some code and avoid this check?

Reply:
Right, the JIT will not optimize this. Yes, this may be better duplicated.

Reply:
done.